### DISCUSSION TITLE
How to properly access device singleton and execute functions?

### DISCUSSION THREAD

**Discussion by onatbas** [koreader/koreader#12392] [Q&A] [answered]
Hi, 

I'm working on a bluetooth page turner plugin, and one of the things I'm trying to do is turn bluetooth on, and start listening to devices. 

My naive attempt at opening the input device as the following failed:

https://github.com/onatbas/bluetooth.koplugin/blob/1d10d332d2c0f5faab976ac2e4b511f5b597388e/main.lua#L37C16-L44C21
```
                callback = function()
                    NetworkMgr:turnOnWifi(function()
				self:onBluetoothOn()
				    local status, err = pcall(function()
				       require("device/kobo/device").input.open("/dev/input/event3")
				    end)                            
				end)
                end,
```

I wasn't too hopeful to begin with as require("device/kobo/device") isn't going to get you the actual instance of the device. My current workaround is to put input.open("/dev/input/event3") directly into device.lua, but that causes me to enforce a KOreader restart and prevents reconnects when bluetooth device connections drop.

So my question is, how do I properly access this running device instance and open the input from my plugin? I'll further develop on that to continuously monitor the existance of the device and attempt re-opens continuously. 

Thank you

**Comment by NiLuJe** [comment, marked as answer]
That would be

```lua
local Device = require("device")

Device.input.open(path)
```

(Except, well, yes, pcall'ed ;p).

c.f., https://github.com/koreader/koreader/blob/master/plugins/externalkeyboard.koplugin for a fancier exemple that supports both scanning input devices on startup and responding to hotplug events.

**Reply by onatbas** [reply]
Fa-antastic. Thank you. I may go for a fancy implementation where I listen for devices as they come up at a later date but since the Input.open maintains an opened device list already, I'll probably just do an infinite loop to try to open it continuously as long as the plugin is activated.
